pkg/apis/core/validation/validation.go

```diff
@@ -2047,21 +2052,31 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 		allErrs = append(allErrs, ValidateImmutableAnnotation(newPvc.ObjectMeta.Annotations[v1.BetaStorageClassAnnotation], oldPvc.ObjectMeta.Annotations[v1.BetaStorageClassAnnotation], v1.BetaStorageClassAnnotation, field.NewPath("metadata"))...)
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes) {
+	if validationOpts.EnableExpansion {
 		// lets make sure storage values are same.
 		if newPvc.Status.Phase == core.ClaimBound && newPvcClone.Spec.Resources.Requests != nil {
 			newPvcClone.Spec.Resources.Requests["storage"] = oldPvc.Spec.Resources.Requests["storage"] // +k8s:verify-mutation:reason=clone
 		}
 
 		oldSize := oldPvc.Spec.Resources.Requests["storage"]
 		newSize := newPvc.Spec.Resources.Requests["storage"]
+		statusSize := oldPvc.Status.Capacity["storage"]
 
 		if !apiequality.Semantic.DeepEqual(newPvcClone.Spec, oldPvcClone.Spec) {
 			specDiff := diff.ObjectDiff(newPvcClone.Spec, oldPvcClone.Spec)
 			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), fmt.Sprintf("spec is immutable after creation except resources.requests for bound claims\n%v", specDiff)))
 		}
 		if newSize.Cmp(oldSize) < 0 {
-			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be less than previous value"))
+			if !validationOpts.EnableRecoverFromExpansionFailure {
+				allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resources", "requests", "storage"), "field can not be less than previous value"))
+			} else {
+				// This validation permits reducing pvc requested size up to capacity recorded in pvc.status
+				// so that users can recover from volume expansion failure, but Kubernetes does not actually
+				// support volume shrinking
+				if newSize.Cmp(statusSize) <= 0 {
```

Comment:
I didn't find this in the KEP. Can you explain why we cannot go back all the way to the original size and why it has to be size + 1?

For this feature, we are already relying on CSI drivers to correctly tell us the actual size of the volume, and also return the proper error codes if an operation is still outstanding.

Reply:
> For this feature, we are already relying on CSI drivers to correctly tell us the actual size of the volume, and also return the proper error codes if an operation is still outstanding.

minor correction - the current design does not depend on getting proper error codes if an operation is still outstanding. Obviously it is nice to have, but it does not depend on it. Giving quota back however does depend on `GetVolume` RPC call returning correct volume size. 

Now coming to original question - I considered various things we can do to go back to original size all the way and I had to reject those options because of underlying complexity.

#### Use a terminal error code from CSI driver if expansion fails
The idea here is - we could use some sort of final error code from CSI driver to indicate expansion has failed permanently and not going to succeed and hence it is safe to go back to original size. The problem is - there aren't any designated final/terminal error codes in CSI. And for expansion operation, it is specially problematic because say you are expanding a PVC backed by a datastore and datastore ran out of disk space but later on some other PVCs got deleted and expansion operation succeeds. In this case PVC will have a final error state but - if we allowed users to go back to original size, then - users will have wrong quota(because expansion controller is continuously retrying expansion). This is of course just an example, but the point is - final error codes aren't reliable for expansion operation.

#### Use returned size from CSI expansion RPC calls as true value
How about some combination of error code and returned size from expansion RPC calls to fix `allocatedResources`? The first problem is - almost every driver I know returns requested size of volume in `ControllerExpandVolume`. For example - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/csi/service/wcp/controller.go#L844 . So if user expands a PVC from 10 to 100GB and that operation succeeded (but we did not record it in pv or pvc.status), and then user shrinks volume back to 20GB and then we call `ControllerExpandVolume` -  they will return 20GB as new size and the second problem is - `NodeExpandVolume` has expanded size as optional return value (Well, it may not be as big of a problem if we come with disclaimer that if you want quota to be replenish you must return correct size in `NodeExpandVolumeResponse`).  

But say - we fixed CSI drivers to return true size of the volume - there is still a problem that, expansion controller does not take any action to expand volumes that have `pvc.Status.Size >= pvc.Spec.Size`. This is especially tricky to flow back onto kubelet because it only tries volume expansion on the node - if and only if `pv.Spec.Size > pvc.Status.Size`, so how do we correct `allocatedResources` for those volumes? Having said that - current design also has the limitation that - to fix `allocatedResources` of PVC - a CSI driver must implement control-plane `GetVolume` RPC call. 

Given all the complexity and practical benefits of going back to original size is almost nothing, I prefer design as proposed.

Reply:
We discussed this during standup. The complication of being able to go back to the original size is mainly due to the current signal that kubelet uses to determine if the controller is finished expanding. kubelet has to start the node expansion after the controller expansion is done, and it [depends on pv size being increased by the controller when finished](https://github.com/kubernetes/kubernetes/blob/c9bff73105dccaabf8a01504fed0dc4fadbaeeda/pkg/volume/util/operationexecutor/operation_generator.go#L1819)

If the size is allowed to go all the way back, then there needs to be a new signal added to indicate that controller expansion is done. We cannot completely use the new `allocatedResources` field as a signal because that will cause the controller and kubelet to race. This will require some additional design to add one more signal and coordinate between the various components.

I think for 1.22 we can continue as planned, but potentially keep the feature in alpha for 1.23 while we consider options for this new signal.